---
title: Özel Durumlar ve Özel Durum Taşıma - C# Programlama Kılavuzu
ms.date: 07/20/2015
helpviewer_keywords:
- exception handling [C#]
- exceptions [C#]
- C# language, exceptions
ms.assetid: 0001887f-4fa2-47e2-8034-2819477e2344
ms.openlocfilehash: b883012cf8f72247ff4e0b47a46eee1854e2d534
ms.sourcegitcommit: 7588136e355e10cbc2582f389c90c127363c02a5
ms.translationtype: MT
ms.contentlocale: tr-TR
ms.lasthandoff: 03/15/2020
ms.locfileid: "76735652"
---
# <a name="exceptions-and-exception-handling-c-programming-guide"></a><span data-ttu-id="dfb25-102">Özel Durumlar ve Özel Durum İşleme (C# Programlama Kılavuzu)</span><span class="sxs-lookup"><span data-stu-id="dfb25-102">Exceptions and Exception Handling (C# Programming Guide)</span></span>

<span data-ttu-id="dfb25-103">C# dilinin özel durum işleme özellikleri, bir program çalışırken meydana gelen beklenmeyen veya istisnai durumlarla başa çıkmanıza yardımcı olur.</span><span class="sxs-lookup"><span data-stu-id="dfb25-103">The C# language's exception handling features help you deal with any unexpected or exceptional situations that occur when a program is running.</span></span> <span data-ttu-id="dfb25-104">Özel durum `try`işleme, başarılı olamayan eylemleri denemek, bunu yapmak için makul olduğuna karar verdiyseniz hataları işlemek ve daha sonra kaynakları temizlemek için , ve `catch` `finally` anahtar kelimeler kullanır.</span><span class="sxs-lookup"><span data-stu-id="dfb25-104">Exception handling uses the `try`, `catch`, and `finally` keywords to try actions that may not succeed, to handle failures when you decide that it is reasonable to do so, and to clean up resources afterward.</span></span> <span data-ttu-id="dfb25-105">Özel durumlar, ortak dil çalışma süresi (CLR), .NET Framework veya herhangi bir üçüncü taraf kitaplıkları veya uygulama kodu tarafından oluşturulabilir.</span><span class="sxs-lookup"><span data-stu-id="dfb25-105">Exceptions can be generated by the common language runtime (CLR), by the .NET Framework or any third-party libraries, or by application code.</span></span> <span data-ttu-id="dfb25-106">Özel durumlar anahtar sözcük `throw` kullanılarak oluşturulur.</span><span class="sxs-lookup"><span data-stu-id="dfb25-106">Exceptions are created by using the `throw` keyword.</span></span>

<span data-ttu-id="dfb25-107">Çoğu durumda, bir özel durum, kodunuzu doğrudan çağırdığı bir yöntem tarafından değil, arama yığınının daha aşağısındabaşka bir yöntem tarafından atılabilir.</span><span class="sxs-lookup"><span data-stu-id="dfb25-107">In many cases, an exception may be thrown not by a method that your code has called directly, but by another method further down in the call stack.</span></span> <span data-ttu-id="dfb25-108">Bu durumda, CLR belirli özel durum türü için bir `catch` blok ile bir yöntem arayarak yığını `catch` gevşetmek ve bulursa ilk tür blok yürütecek.</span><span class="sxs-lookup"><span data-stu-id="dfb25-108">When this happens, the CLR will unwind the stack, looking for a method with a `catch` block for the specific exception type, and it will execute the first such `catch` block that if finds.</span></span> <span data-ttu-id="dfb25-109">Arama yığınının `catch` herhangi bir yerinde uygun bir engel bulamazsa, işlemi sonlandırır ve kullanıcıya bir ileti görüntüler.</span><span class="sxs-lookup"><span data-stu-id="dfb25-109">If it finds no appropriate `catch` block anywhere in the call stack, it will terminate the process and display a message to the user.</span></span>

<span data-ttu-id="dfb25-110">Bu örnekte, bir yöntem sıfıra bölme için test ve hata yakalar.</span><span class="sxs-lookup"><span data-stu-id="dfb25-110">In this example, a method tests for division by zero and catches the error.</span></span> <span data-ttu-id="dfb25-111">Özel durum işleme olmadan, bu program **bir DivideByZeroException ile sonlandırılmaz işlenmemiş** hata oldu.</span><span class="sxs-lookup"><span data-stu-id="dfb25-111">Without the exception handling, this program would terminate with a **DivideByZeroException was unhandled** error.</span></span>

[!code-csharp[csProgGuideExceptions#18](~/samples/snippets/csharp/VS_Snippets_VBCSharp/csProgGuideExceptions/CS/Exceptions.cs#18)]

## <a name="exceptions-overview"></a><span data-ttu-id="dfb25-112">Özel Durumlara Genel Bakış</span><span class="sxs-lookup"><span data-stu-id="dfb25-112">Exceptions Overview</span></span>

<span data-ttu-id="dfb25-113">Özel durumlar aşağıdaki özelliklere sahiptir:</span><span class="sxs-lookup"><span data-stu-id="dfb25-113">Exceptions have the following properties:</span></span>

- <span data-ttu-id="dfb25-114">Özel durumlar, sonuçta tüm türlerden `System.Exception`türetilmiştir.</span><span class="sxs-lookup"><span data-stu-id="dfb25-114">Exceptions are types that all ultimately derive from `System.Exception`.</span></span>
- <span data-ttu-id="dfb25-115">Özel `try` durumlar atabilecek deyimlerin etrafında bir blok kullanın.</span><span class="sxs-lookup"><span data-stu-id="dfb25-115">Use a `try` block around the statements that might throw exceptions.</span></span>
- <span data-ttu-id="dfb25-116">`try` Blokta bir özel durum oluştuğunda, denetim akışı çağrı yığınının herhangi bir yerinde bulunan ilk ilişkili özel durum işleyicisine atlar.</span><span class="sxs-lookup"><span data-stu-id="dfb25-116">Once an exception occurs in the `try` block, the flow of control jumps to the first associated exception handler that is present anywhere in the call stack.</span></span> <span data-ttu-id="dfb25-117">`catch` C#'da, anahtar kelime bir özel durum işleyicisi tanımlamak için kullanılır.</span><span class="sxs-lookup"><span data-stu-id="dfb25-117">In C#, the `catch` keyword is used to define an exception handler.</span></span>
- <span data-ttu-id="dfb25-118">Belirli bir özel durum için özel durum işleyicisi yoksa, program bir hata iletisi ile yürütmeyi durdurur.</span><span class="sxs-lookup"><span data-stu-id="dfb25-118">If no exception handler for a given exception is present, the program stops executing with an error message.</span></span>
- <span data-ttu-id="dfb25-119">İşleyebilir ve uygulamayı bilinen bir durumda bırakmadığınız sürece bir özel durum yakalamayın.</span><span class="sxs-lookup"><span data-stu-id="dfb25-119">Do not catch an exception unless you can handle it and leave the application in a known state.</span></span> <span data-ttu-id="dfb25-120">Yakalarsanız, `System.Exception` `catch` bloğun sonundaki anahtar `throw` kelimeyi kullanarak yeniden atayın.</span><span class="sxs-lookup"><span data-stu-id="dfb25-120">If you catch `System.Exception`, re-throw it using the `throw` keyword at the end of the `catch` block.</span></span>
- <span data-ttu-id="dfb25-121">Bir `catch` blok bir özel durum değişkeni tanımlıyorsa, oluşan özel durum türü hakkında daha fazla bilgi edinmek için kullanabilirsiniz.</span><span class="sxs-lookup"><span data-stu-id="dfb25-121">If a `catch` block defines an exception variable, you can use it to obtain more information about the type of exception that occurred.</span></span>
- <span data-ttu-id="dfb25-122">Özel `throw` durumlar, anahtar kelime kullanılarak bir program tarafından açıkça oluşturulabilir.</span><span class="sxs-lookup"><span data-stu-id="dfb25-122">Exceptions can be explicitly generated by a program by using the `throw` keyword.</span></span>
- <span data-ttu-id="dfb25-123">Özel durum nesneleri, arama yığınının durumu ve hatanın metin açıklaması gibi hata hakkında ayrıntılı bilgiler içerir.</span><span class="sxs-lookup"><span data-stu-id="dfb25-123">Exception objects contain detailed information about the error, such as the state of the call stack and a text description of the error.</span></span>
- <span data-ttu-id="dfb25-124">Bir özel `finally` durum atılsa bile bloktaki kod yürütülür.</span><span class="sxs-lookup"><span data-stu-id="dfb25-124">Code in a `finally` block is executed even if an exception is thrown.</span></span> <span data-ttu-id="dfb25-125">Örneğin, `finally` `try` blokta açılan akışları veya dosyaları kapatmak için kaynakları serbest bırakmak için bir blok kullanın.</span><span class="sxs-lookup"><span data-stu-id="dfb25-125">Use a `finally` block to release resources, for example to close any streams or files that were opened in the `try` block.</span></span>
- <span data-ttu-id="dfb25-126">.NET Framework'de yönetilen özel durumlar Win32 yapılandırılmış özel durum işleme mekanizmasının üstüne uygulanır.</span><span class="sxs-lookup"><span data-stu-id="dfb25-126">Managed exceptions in the .NET Framework are implemented on top of the Win32 structured exception handling mechanism.</span></span> <span data-ttu-id="dfb25-127">Daha fazla bilgi için [Win32 Yapılandırılmış Özel Durum İşleme'nin Derinliklerinde](http://bytepointer.com/resources/pietrek_crash_course_depths_of_win32_seh.htm) [Yapılandırılmış Özel Durum İşleme (C/C++)](/cpp/cpp/structured-exception-handling-c-cpp) ve Bir Kilitlenme Kursu'na bakın.</span><span class="sxs-lookup"><span data-stu-id="dfb25-127">For more information, see [Structured Exception Handling (C/C++)](/cpp/cpp/structured-exception-handling-c-cpp) and [A Crash Course on the Depths of Win32 Structured Exception Handling](http://bytepointer.com/resources/pietrek_crash_course_depths_of_win32_seh.htm).</span></span>

## <a name="related-sections"></a><span data-ttu-id="dfb25-128">İlgili Bölümler</span><span class="sxs-lookup"><span data-stu-id="dfb25-128">Related Sections</span></span>

<span data-ttu-id="dfb25-129">Özel durumlar ve özel durum işleme hakkında daha fazla bilgi için aşağıdaki makalelere bakın:</span><span class="sxs-lookup"><span data-stu-id="dfb25-129">See the following articles for more information about exceptions and exception handling:</span></span>

- [<span data-ttu-id="dfb25-130">Özel Durumları Kullanma</span><span class="sxs-lookup"><span data-stu-id="dfb25-130">Using Exceptions</span></span>](using-exceptions.md)
- [<span data-ttu-id="dfb25-131">Özel Durum Taşıma</span><span class="sxs-lookup"><span data-stu-id="dfb25-131">Exception Handling</span></span>](exception-handling.md)
- [<span data-ttu-id="dfb25-132">Özel Durumlar Oluşturma ve Atma</span><span class="sxs-lookup"><span data-stu-id="dfb25-132">Creating and Throwing Exceptions</span></span>](creating-and-throwing-exceptions.md)
- [<span data-ttu-id="dfb25-133">Derleyicinin Ürettiği Özel Durumlar</span><span class="sxs-lookup"><span data-stu-id="dfb25-133">Compiler-Generated Exceptions</span></span>](compiler-generated-exceptions.md)
- [<span data-ttu-id="dfb25-134">Try/catch (C# Programlama Kılavuzu) kullanarak bir özel durum nasıl işleyebilir</span><span class="sxs-lookup"><span data-stu-id="dfb25-134">How to handle an exception using try/catch (C# Programming Guide)</span></span>](how-to-handle-an-exception-using-try-catch.md)
- [<span data-ttu-id="dfb25-135">Finally anahtar sözcüğünü kullanarak temizleme kodu yürütme</span><span class="sxs-lookup"><span data-stu-id="dfb25-135">How to execute cleanup code using finally</span></span>](how-to-execute-cleanup-code-using-finally.md)
- [<span data-ttu-id="dfb25-136">CLS dışı özel durumu yakalama</span><span class="sxs-lookup"><span data-stu-id="dfb25-136">How to catch a non-CLS exception</span></span>](how-to-catch-a-non-cls-exception.md)

## <a name="c-language-specification"></a><span data-ttu-id="dfb25-137">C# Dil Belirtimi</span><span class="sxs-lookup"><span data-stu-id="dfb25-137">C# Language Specification</span></span>

<span data-ttu-id="dfb25-138">Daha fazla bilgi için [C# Dil Belirtiminde](/dotnet/csharp/language-reference/language-specification/introduction) [Özel Durumlar'a](~/_csharplang/spec/exceptions.md) bakın.</span><span class="sxs-lookup"><span data-stu-id="dfb25-138">For more information, see [Exceptions](~/_csharplang/spec/exceptions.md) in the [C# Language Specification](/dotnet/csharp/language-reference/language-specification/introduction).</span></span> <span data-ttu-id="dfb25-139">Dil belirtimi, C# sözdizimi ve kullanımı için kesin bir kaynaktır.</span><span class="sxs-lookup"><span data-stu-id="dfb25-139">The language specification is the definitive source for C# syntax and usage.</span></span>

## <a name="see-also"></a><span data-ttu-id="dfb25-140">Ayrıca bkz.</span><span class="sxs-lookup"><span data-stu-id="dfb25-140">See also</span></span>

- <xref:System.SystemException>
- [<span data-ttu-id="dfb25-141">C# Programlama Kılavuzu</span><span class="sxs-lookup"><span data-stu-id="dfb25-141">C# Programming Guide</span></span>](../index.md)
- [<span data-ttu-id="dfb25-142">C# Anahtar Kelimeler</span><span class="sxs-lookup"><span data-stu-id="dfb25-142">C# Keywords</span></span>](../../language-reference/keywords/index.md)
- [<span data-ttu-id="dfb25-143">throw</span><span class="sxs-lookup"><span data-stu-id="dfb25-143">throw</span></span>](../../language-reference/keywords/throw.md)
- [<span data-ttu-id="dfb25-144">try-catch</span><span class="sxs-lookup"><span data-stu-id="dfb25-144">try-catch</span></span>](../../language-reference/keywords/try-catch.md)
- [<span data-ttu-id="dfb25-145">try-finally</span><span class="sxs-lookup"><span data-stu-id="dfb25-145">try-finally</span></span>](../../language-reference/keywords/try-finally.md)
- [<span data-ttu-id="dfb25-146">try-catch-finally</span><span class="sxs-lookup"><span data-stu-id="dfb25-146">try-catch-finally</span></span>](../../language-reference/keywords/try-catch-finally.md)
- [<span data-ttu-id="dfb25-147">Özel durumlar</span><span class="sxs-lookup"><span data-stu-id="dfb25-147">Exceptions</span></span>](../../../standard/exceptions/index.md)
